row_buffer.go

```diff
@@ -0,0 +1,505 @@
+				maxRepetitionLevel: byte(leafColumn.MaxRepetitionLevel),
+				maxDefinitionLevel: byte(leafColumn.MaxDefinitionLevel),
+			},
+		}
+	}
+
+	columnChunks := make([]ColumnChunk, len(chunks))
+	for i := range chunks {
+		columnChunks[i] = &chunks[i]
+	}
+	return columnChunks
+}
+
+// SortingColumns returns the list of columns that rows are expected to be
+// sorted by.
+//
+// The list of sorting columns is configured when the buffer is created and used
+// when it is sorted.
+//
+// Note that unless the buffer is explicitly sorted, there are no guarantees
+// that the rows it contains will be in the order specified by the sorting
+// columns.
+func (buf *RowBuffer[T]) SortingColumns() []SortingColumn { return buf.sorting }
+
+// Schema returns the schema of rows in the buffer.
+func (buf *RowBuffer[T]) Schema() *Schema { return buf.schema }
+
+// Len returns the number of rows in the buffer.
+//
+// The method contributes to satisfying sort.Interface.
+func (buf *RowBuffer[T]) Len() int { return buf.numRows }
+
+// Less compares the rows at index i and j according to the sorting columns
+// configured on the buffer.
+//
+// The method contributes to satisfying sort.Interface.
+func (buf *RowBuffer[T]) Less(i, j int) bool {
+	return buf.compare(buf.rows[i], buf.rows[j]) < 0
+}
+
+// Swap exchanges the rows at index i and j in the buffer.
+//
+// The method contributes to satisfying sort.Interface.
+func (buf *RowBuffer[T]) Swap(i, j int) {
+	buf.rows[i], buf.rows[j] = buf.rows[j], buf.rows[i]
+}
+
+// Rows returns a Rows instance exposing rows stored in the buffer.
+//
+// The rows returned are a snapshot at the time the method is called.
+// The returned rows and values read from it remain valid until the next call
+// to Reset on the buffer.
+func (buf *RowBuffer[T]) Rows() Rows {
+	return &rowBufferRows{rows: buf.rows[:buf.numRows], schema: buf.schema}
+}
+
+// Write writes rows to the buffer, returning the number of rows written.
+func (buf *RowBuffer[T]) Write(rows []T) (int, error) {
+	buf.reserve(len(rows))
+
+	for i := range rows {
+		bufRow := buf.rows[buf.numRows][:0]
+		bufRow = buf.schema.Deconstruct(bufRow, &rows[i])
+		buf.capture(bufRow)
+		buf.rows[buf.numRows] = bufRow
+		buf.numRows++
+	}
+
+	return len(rows), nil
+}
+
+// WriteRows writes parquet rows to the buffer, returing the number of rows
+// written.
+func (buf *RowBuffer[T]) WriteRows(rows []Row) (int, error) {
+	buf.reserve(len(rows))
+
+	for _, row := range rows {
+		bufRow := buf.rows[buf.numRows][:0]
+		bufRow = append(bufRow, row...)
+		buf.capture(bufRow)
+		buf.rows[buf.numRows] = bufRow
+		buf.numRows++
+	}
+
+	return len(rows), nil
+}
+
+func (buf *RowBuffer[T]) copyBytes(v []byte) []byte {
+	if free := cap(buf.buffer) - len(buf.buffer); free < len(v) {
+		newCap := 2 * cap(buf.buffer)
+		if newCap == 0 {
+			newCap = defaultReadBufferSize
+		}
+		for newCap < len(v) {
+			newCap *= 2
+		}
+		buf.buffer = make([]byte, 0, newCap)
```

Comment:
Don't you need to copy the existing values here or am I nuts

Reply:
No need, the previous rows will reference the old buffer and GC will reclaim memory when the pointers are cleared.